### DISCUSSION TITLE
slowdown

### DISCUSSION THREAD

**Discussion by ghajini** [uBlockOrigin/uBlock-issues#2076] [Q&A]
iam experiencing this slowdown by ublock origin(as per firefox dialog)

ubo 1.42.1b2 ,firefox 98.0.2

 **Website=** `moviewatchonline.com.pk`

[2Firefox 2022-04-03 14.32 profile.json.gz](https://github.com/uBlockOrigin/uBlock-issues/files/8404145/2Firefox.2022-04-03.14.32.profile.json.gz)
[Firefox 2022-04-03 14.22 profile.json.gz](https://github.com/uBlockOrigin/uBlock-issues/files/8404146/Firefox.2022-04-03.14.22.profile.json.gz)

![Screenshot (306)](https://user-images.githubusercontent.com/20338483/161420470-f378b1e3-d698-42c2-8308-65a95fc0bbd8.png)

**Comment by gorhill** [comment]
I can reproduce with Firefox, not with Chromium.

It appears Firefox's `Element.classList.values()` is rather slow compare to Chromium. Most of the time it's probably unnoticeable but this page really puts the API to test. Using `view-source:https://moviewatchonline.com.pk/`, I find that there are 40 `article` elements, each with something like ~1,500 class tokens, and such a high number of distinct class tokens is what causes `Element.classList.values()` to be slow, but only in Firefox.

I might want to dig further as this might be something Firefox devs could want to address, surely `Element.classList` shouldn't be that much slower in Firefox than in Chromium. Regarding uBO though, there is no trivial fix other than disabling generic cosmetic filters for now on this site, `@@||moviewatchonline.com.pk^$ghide`.

I might want to experiment with having uBO parsing the class tokens and not rely on `Element.classList.values()`.

**Comment by gorhill** [comment]
> I might want to experiment with having uBO parsing the class tokens and not rely on Element.classList.values()

Rewriting `classesFromNode()` to not rely on `Element.classList` leads to significant performance improvement on that page in Firefox. For Chromium there is a marginal improvement (though probably unnoticeable by end users), so I think I will go with the rewrite.

We will have to watch for regressions, as I was relying on `Element.classList` to return me properly parsed class names, but [as per specs](https://dom.spec.whatwg.org/#concept-ordered-set-parser), it turns out the parsing is simpler than I thought so that doing it manually might work fine. uBO does not need the class names to be processed as detailed in the spec, it just needs to get them all as fast as possible.

**Reply by ghajini** [reply]
i tried ubo 1.42.3b0 ,firefox 98.0.2
much improvement ,though there is sometimes lag ;no more page slow down message by firefox,here is profiling data of interest to you
[Firefox 2022-04-04 00.04 profile.json.gz](https://github.com/uBlockOrigin/uBlock-issues/files/8405256/Firefox.2022-04-04.00.04.profile.json.gz)

**Reply by gorhill** [reply]
Your profile does not match 1.42.3b0, it's a profile based on uBO 1.42.2 or earlier.

**Reply by ghajini** [reply]
where can i share profiling data ,github isn't allowing me to upload

**Reply by ghajini** [reply]
i have filed bug  anyway https://bugzilla.mozilla.org/show_bug.cgi?id=1762796

**Reply by gorhill** [reply]
> Your profile does not match 1.42.3b0, it's a profile based on uBO 1.42.2 or earlier.

Sorry, I was mistaken, I tried again to load it in Nightly, and I see different results, I think I must have loaded the wrong profile.

uBO is not the biggest issue performance-wise as per profiling results -- the page own's code is what is occupying the CPU.

**Reply by gorhill** [reply]
> i have filed bug anyway https://bugzilla.mozilla.org/show_bug.cgi?id=1762796

I didn't expect such a fast fix. When it's available on Nightly, I will compare which approach is best performance-wise and maybe restore the previous code. I realize that currently with the new code there would be an issue if there were whitespaces other than space character -- there had been cases of newlines being used inside attribute values, this would cause an issue with the new code.

**Reply by ghost** [reply]
This bug report is now closed as fixed in Firefox Nightly 101. The Firefox Nightly 101 [Bugzilla List](https://bugzilla.mozilla.org/buglist.cgi?j_top=OR&f1=target_milestone&o3=equals&v3=Firefox%20101&o1=equals&resolution=FIXED&o2=anyexact&query_format=advanced&f3=target_milestone&f2=cf_status_firefox101&bug_status=RESOLVED&bug_status=VERIFIED&bug_status=CLOSED&v1=mozilla101&v2=fixed%2Cverified&limit=0) of fixes also shows it listed. Firefox Nightly 101 is out now as well.

**Reply by gorhill** [reply]
> I realize that currently with the new code there would be an issue if there were whitespaces other than space character -- there had been cases of newlines being used inside attribute values

I forgot to revisit after fix in Firefox and this happened: <https://www.reddit.com/r/uBlockOrigin/comments/zurhqa/general_rule_does_not_work_on_first_visit/>.

**Comment by mayankleoboy** [comment]
This is being improved here: https://bugzilla.mozilla.org/show_bug.cgi?id=1762796

**Reply by gettysburg** [reply]
Indeed, thank @ghajini for filing the bug report 😃
